File: algorithms/[1319]连通网络的操作次数/solution.py
# topics = ["图", "广度优先搜索, "深度优先搜索"]

from collections import deque
from typing import List, Set


class Solution:
    def buildGraph(self, n: int, edges: List[List[int]]) -> List[Set[int]]:
        """
        无向图-邻接表
        """
        graph: List[Set[int]] = [set() for _ in range(n)]

        for i, j in edges:
            graph[i].add(j)
            graph[j].add(i)

        return graph

    def bfs(self, idx: int, visited: List[int], graph: List[Set[int]]) -> None:
        visited[idx] = 1
        queue = deque(graph[idx])

        while queue:
            cur = queue.popleft()
            if not visited[cur]:
                visited[cur] = 1
                queue.extend(graph[cur])

    def dfs(self, idx: int, visited: List[int], graph: List[Set[int]]) -> None:
        visited[idx] = 1
        for node in graph[idx]:
            if not visited[node]:
                self.dfs(node, visited, graph)

    def makeConnected(self, n: int, connections: List[List[int]]) -> int:
        """
        BFS / DFS
        time O(m + n), space O(m + n), m 为 connections 的长度
        """
        if len(connections) < n - 1:
            return -1

        graph = self.buildGraph(n, connections)

        # 连通分量数
        res = 0
        # 记录已遍历过的顶点
        visited = [0] * n

        for i in range(n):
            if not visited[i]:
                # 解法一 bfs
                # self.bfs(i, visited, graph)
                # 解法二 dfs
                self.dfs(i, visited, graph)
                res += 1

        return res - 1

File: algorithms/[1319]连通网络的操作次数/solution.ts
// topics = ["图", "并查集"]

export class UnionFind {
  parent: number[];
  rank: number[];
  setCount: number;
  constructor(n: number) {
    this.parent = Array.from({ length: n }, (_, i) => i);
    this.rank = Array.from({ length: n }, (_, i) => 1);
    this.setCount = n;
  }
  union(i: number, j: number): void {
    const x = this.find(i);
    const y = this.find(j);

    if (x === y) return;

    if (this.rank[x] <= this.rank[y]) {
      this.parent[x] = y;
    } else {
      this.parent[y] = x;
    }

    if (this.rank[x] === this.rank[y]) {
      this.rank[y] + 1;
    }

    this.setCount -= 1;
  }
  find(x: number): number {
    if (this.parent[x] !== x) {
      this.parent[x] = this.find(this.parent[x]);
    }

    return this.parent[x];
  }
}

/**
 * 并查集求连通分量
 * time O(m·α(n)), space O(n), m 为 connections 的长度, α 是阿克曼函数的反函数
 */
function makeConnected(n: number, connections: number[][]): number {
  const len = connections.length;

  // 当计算机的数量为 n 时，至少需要 n-1 根线才能将它们进行连接
  if (len < n - 1) return -1;

  // 已知边集合，求连通分量 k
  // 移动 k - 1 根线即可将所有计算机连接
  const uf = new UnionFind(n);
  connections.forEach(([i, j]) => uf.union(i, j));
  return uf.setCount - 1;
}
